src/binaryServerStarter.ts

```diff
@@ -0,0 +1,262 @@
+  let binaryPath: string = config.get("server.binary.path");
+  if (binaryPath) {
+    if (fs.existsSync(binaryPath)) {
+      return Promise.resolve(binaryPath);
+    }
+    window.showErrorMessage('The specified XML language server binary could not be found. Using the default binary...');
+  }
+  let server_home: string = path.resolve(__dirname, '../server');
+  let binaries: Array<string> = glob.sync(`**/${getServerBinaryNameWithoutExtension()}*`, { cwd: server_home });
+  binaries = binaries.filter((path) => { return !JAR_ZIP_AND_HASH_REJECTOR.test(path) });
+  if (binaries.length) {
+    return Promise.resolve(path.resolve(server_home, binaries[0]));
+  }
+  // Download it, then return the downloaded binary's location
+  return downloadBinary();
+}
+
+/**
+ * Downloads LemMinX binary under the `server` directory and returns the path to the binary as a Promise
+ *
+ * @returns The path to the LemMinX binary
+ * @throws If the LemMinX binary download fails
+ */
+async function downloadBinary(): Promise<string> {
+  window.setStatusBarMessage('Downloading XML language server binary...', 2000);
+  return new Promise((resolve, reject) => {
+    const handleResponse = (response: http.IncomingMessage) => {
+      const statusCode = response.statusCode;
+      if (statusCode === 303) {
+        httpHttpsGet(response.headers.location, handleResponse)
+          .on('error', (e) => {
+            reject(`Server binary download failed: ${e.message}`);
+          });
+      } else if (statusCode === 200) {
+        if (/^application\/zip$/.test(response.headers['content-type'])) {
+          acceptZipDownloadResponse(response).then(resolve, reject);
+        } else {
+          // Not a ZIP, assume its a binary
+          acceptBinaryDownloadResponse(response).then(resolve, reject);
+        }
+      } else {
+        reject(`Server binary download failed: status code ${statusCode}`);
+      }
+    }
+    httpHttpsGet(`${getServerBinaryDownloadUrl()}`, handleResponse)
+      .on('error', (e) => {
+        reject(`Server binary download failed: ${e.message}`);
+      });
+  });
+}
+
+/**
+ * Returns true if the hash of the binary matches the expected hash and false otherwise
+ *
+ * @param binaryPath the path to the binary to check
+ * @returns true if the hash of the binary matches the expected hash and false otherwise
+ */
+async function checkBinaryHash(binaryPath: string): Promise<boolean> {
+  const hash: Hash = createHash('sha256');
+  return new Promise((resolve, reject) => {
+    fs.readFile(path.resolve(binaryPath), (err, fileContents) => {
+      if (err) {
+        reject(err)
+      };
+      resolve(fileContents);
+    });
+  })
+    .then((fileContents: string) => {
+      hash.update(fileContents);
+      const hashDigest: string = hash.digest('hex').toLowerCase();
+      const expectedHashPath: string = path.resolve(__dirname, '../server', `${getServerBinaryNameWithoutExtension()}.sha256`);
```

Comment:
we shouldn't assume the binary lives under ../server. If the path is provided from the `xml.server.binary.path` setting, then we search sha256 file under the same directory

Reply:
I check under the same directory first, then if its not there I use the built-in hash file